src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs

```diff
@@ -19,7 +20,7 @@ public class AttributeCollection : ICollection, IEnumerable
         /// </summary>
         public static readonly AttributeCollection Empty = new AttributeCollection(null);
 
-        private static Hashtable? s_defaultAttributes;
+        private static Dictionary<Type, Attribute?>? s_defaultAttributes;
```

Comment:
One reason we haven't mass replaced `Hashtable` with `Dictionary<K,V>` is because the thread-safety differences between the two implementations.

See https://stackoverflow.com/questions/10255456/difference-between-dictionary-and-hashtable-in-terms-of-thread-safety

Reply:
Good point! I recently learned about the difference in thread safety for the two types, but hadn't considered the implications until you mentioned it.

I'll take another look through all the changes with this in mind. Are there any specific guidelines you'd like me to follow? For instance, do we want to aim to remove all the non-generic uses (like the issue says), or instead is the issue overly optimistic, and in multi-threaded scenarios we should instead leave the Hashtables as-is and avoid the churn?

Your thoughts are appreciated. Thanks!

Reply:
> in multi-threaded scenarios we should instead leave the Hashtables as-is and avoid the churn?

If the Hashtable is ever read outside of a lock but updated inside of a lock, it should remain a Hashtable or a ConcurrentDictionary.

Reply:
Another difference to keep in mind is if anything from the collection is ever exposed, e.g. an enumerator.  We've had accidental breaking changes in the past where a Hashtable was changed to a Dictionary and that in turn changed the type of the values yielded from an enumerator.

Reply:
Thanks for the info! I've gone back with an eye towards concurrent use, and also did find a case where we expose an enumerator, so I've reverted those changes as well.